### DISCUSSION TITLE
Various options when Compiling with Tasmocompiler and their size

### DISCUSSION THREAD

**Discussion by NsinghP** [arendst/Tasmota#12379] [Q&A] [answered]
When Compiling with newest Tasmocompiler @benzino77 ‘s site and using Wemos/NodeMCU option, we are presented with following options:

(1)Air/Gas Sensor (2) Amazon Alexa, (3) Arduino Slave, (4) Display Vcc, (5) Displays (12C/SPI) (6) Distance Sensors, (7) Domoticz, (8)Energy sensors, (9) Home Assistant, (10) IO port expander, (11) IR Support, (12) KNX, (13) Light sensors, (14) mDNS discovery, (15) MQTT over TLS, (16) RF transceiver, (17) Rules, (18) Script,  (19) SD card/LittleFS, (20) Temp/Hum sensor, (21) Timers, (22) Tuya MCU, (23) web interface, (24) WS2812 Leds, (25) Zigbee.
Out of these 4 are already selected by default as follows:
(17) Rules, (19) SD card/LittleFS, (21) Timers, (23) Web Interface

In addition to above four already selected by default I am using two more as follows:
(7) Domoticz, (15) MQTT over TLS
Bringing the options selected by me to a total of six. Resulting .bin file is working good and there are no problems when I upgrade older versions of tasmota with this new .bin file using firmware.bin.gz version which is 452 KB in size.

My questions are:

1)	Can I keep adding more options above and beyond the six I am already using without any problem. What is the max size that .bin.gz file can be without running into problems flashing/upgrading. I am using ESP8266 12F based device with 4MB flash. 
2)	It would be nice to read up on what all these option do. Is there a link/site which covers explanation for all the options in one place?
3)	It would also be nice to see right after the option name how much size it is so it is easy to calculate when you have reached max size for your .bin file. Can this be done?

Regards,

Navi

**Comment by ascillato** [comment]
Hi,

1, yes you can up to the file size got bigger than what the hardware supports.
In the case of the ESP8266, no matter the size of your flash, the MCU can only run programs up to 1Mbytes. So, with the settings area+SDK settings, you have a bit more of 900Kbytes for Tasmota. The rest of the flash can be used for store data, not for programs.
Then, for OTA, you also need extra space too in the program area. In the actual memory map of Tasmota, OTA is placed also in the same 1MByte. So if you have a bin of 520K that you want to updload by OTA and your actual firmware in the device is 340K, the OTA process will work just fine because you have about 600K free space.
If your firmware in the device have already a 520K firmware and you want to upload another of 520K, you need to upload first a smaller firmware (Tasmota-minimal) as an intermediate step to free space, and then the new 520 firmware. 
With the 1MB memory map, if you want to have in the device a firmware of about 900KB, you will need to flash by wire. Not by OTA.

There is another option to use a another memory map linker of 1mb for program and another 1mb for OTA and the rest for data (that option is the maximum supported by the ESP8266 but requires a flash of 2MB or more)
The maximum flash supported by the ESP8266 is 16MB, where 1MB is program, another 1MB is OTA and 14MB are available for data storage.
Using this other memory map linker (only usable for devices with 2mb or more) you can have a 900KB firmware and do an OTA upgrade with a 900KB firmware too.

2- and 3- should be asked in TasmoCompiler repository directly.

**Reply by NsinghP** [reply]
Thanks, also posted on tasmocompiler page. please see https://github.com/benzino77/tasmocompiler/discussions/208

**Reply by NsinghP** [reply]
> Using this other memory map linker (only usable for devices with 2mb or more) you can have a 900KB firmware and do an OTA upgrade with a 900KB firmware too.


I think my flash fits this description, and I should be able to safely do 900KB:

Flash Size	4096 kB
Program Flash Size	4096 kB
Program Size	630 kB
Free Program Space	1416 kB
Free Memory	19.8 kB

By the way what is this 19.8 kb free memory it is referring to?

**Reply by sfromis** [reply]
Free memory is RAM. Enabling various features will reduce available RAM for heap, and Tasmota may get unstable if the number gets too low. IIRC, 19.8kB is not danger zone, but still lower than the "usual" 26-28 kB on ESP8266.

**Reply by NsinghP** [reply]
> 19.8kB is not danger zone, but still lower than the "usual" 26-28 kB on ESP8266.

Thanks, good too know, now I will not get carried away with loading unnecessary features, I am surprised that 6 features:
 (1) Rules, (2) SD card/LittleFS, (3) Timers, (4) Web Interface, (5) Domoticz, (6) MQTT over TLS
are using up so much RAM, I am sure it has something to do with MQTT over TLS feature.

**Reply by s-hadinger** [reply]
As described here https://tasmota.github.io/docs/TLS/

TLS takes 7kb of ram once connected.

**Reply by NsinghP** [reply]
> 19.8kB is not danger zone, but still lower than the "usual" 26-28 kB on ESP8266.

Just Did an OTA firmware upgrade:
After OTA upgrade to 9.5.0

Flash Size	4096 kB
Program Flash Size	1024 kB
Program Size	607 kB
Free Program Space	396 kB
Free Memory	27.9 kB

"Consoles" option  changed into "Console" option and file management system was missing. Through web console I gave the command Reset 1 to see if file management system comes back. But After this Reset-1 device seemed to be bricked. Tried Button-1 reset for 40 seconds. Nothing happened. Did the device recovery sequence (On off Six times), Now I see tasmota AP. (but feel this is undesirable behavior as I have lost all changes that were implemented in my custom 9.4.0.5 .bin device. which included a different AP name and passphrase). and it looks like I have also lost the MQTT-TLS feature I had compiled with my 9.4.0.5 device.

Is this expected after an OTA upgrade initiated from a custom .bin file flashed device?

Edit: I just noticed that Program Flash Size also changed to 1024kb from 4096kb before. I think that may have something to do with bricking symptoms displayed by my device.

**Reply by sfromis** [reply]
If you upgrade to binary with an incompatible memory model, you're screwed. Always keep same model, or erase/flash. Sometimes it may work enough to allow you to do a recovery, but that's certainly not something you should expect.

**Reply by NsinghP** [reply]
> If you upgrade to binary with an incompatible memory model, you're screwed. Always keep same model, or erase/flash. Sometimes it may work enough to allow you to do a recovery, but that's certainly not something you should expect.

Ha ha, does this mean that I should disable OTA upgrades from my device? or is it possible to select memory model (i.e. 4MB chip/nodemcu etc.) when doing OTA upgrade?

Edit: Is it even possible to disable OTA upgrade when compiling on tasmocompiler?

**Reply by sfromis** [reply]
No, you "should" not disable OTA upgrade, but merely be careful what you upgrade to when going outside of the "most common" setup, which for ESP8266 is 1MB. Memory model is selected when building the binary, typically by selecting a board variant in [platformio_override.ini](https://github.com/arendst/Tasmota/blob/development/platformio_override_sample.ini)

**Reply by NsinghP** [reply]
> No, you "should" not disable OTA upgrade, but merely be careful what you upgrade to when going outside of the "most common" setup, which for ESP8266 is 1MB.

But if, like in my case, i have compiled on tasmocompiler choosing the option wemos/nodemcu, it may be dangerous to keep OTA active, as OTA by default is designed for 1MB memory model. If OTA upgrade is done by mistake it will 100% brick or Semi Brick the device. Either this or select Generic option to limit my binary to 1MB memory model when compiling on tasmocompiler.  It does not make sense to keep OTA active for 4MB devices that have been compiled to use anything else than 1MB memory model.

Edit: From end-user friendliness perspective I think it is desirable to at least have a warning on OTA page about not upgrading OTA if using 4MB memory model or redesigning the OTA page so that it identifies if the device is using 4MB memory model and changes the OTA url to point to a 4MB memory model version of the upgrade.  I don't know how difficult it is to implement the later but it can prevent many bricked devices, where the end-user has no idea of what he/she is about to encounter after hitting the innocent looking upgrade button.

**Reply by Jason2866** [reply]
Upgrading between different memory models WAS a problem in the past.
Since 12th January it is possible. As explained https://github.com/arendst/Tasmota/blob/development/platformio_override_sample.ini#L54-L63
This does mean you can upgrade every device via OTA with a different memory model when it is based on a commit after 12.01.2021

**Reply by sfromis** [reply]
No, OTA is not designed for any specific memory model, it is always whatever was used for the binary you select when upgrading. Anyway if all binaries are new enough, it may be a non-issue now.

More generally, the design of Tasmota does not include protecting the user from doing bad things.

**Reply by NsinghP** [reply]
> Anyway if all binaries are new enough, it may be a non-issue now.

Thanks, Let me test as new versions become available. In current case I was upgrading from 9.4.0.5 to 9.5, so the problem should not have happened as they are both latest binaries. So I wonder what caused my problem? Could it be some of the features in my custom strings? could it be select able options on tasmocompiler page? or something else. I will keep an eye out for anything out of the ordinary and post here.

**Reply by NsinghP** [reply]
> This does mean you can upgrade every device via OTA with a different memory model when it is based on a commit after 12.01.2021

I wonder if I ran into a problem because my binary is compiled on tasmocompiler, maybe the changes you refer to in your link are not yet implemented there? so my binary was incompatible?

**Reply by NsinghP** [reply]
I was able to recreate the problem:

Did serial erase flash of the chip using Tasmotizer and reverted back to Custom Tasmocompiler compled Tasmota 9.4.05 bin file. After booting into this device, started OTA upgrade. Device was successfully flashed to Tasmota 9.5 OTA. Again File management is gone, MQTT-TLS is gone, file system has been set by the OTA upgrade to 1MB model. Could this be a bug or just a faulty chip? Here are some pictues:

![Screenshot (190)](https://user-images.githubusercontent.com/69470381/122429878-bef73d80-cfb0-11eb-9efe-5f3db40066f0.png)
![Screenshot (191)](https://user-images.githubusercontent.com/69470381/122429933-cc142c80-cfb0-11eb-91c6-8b0cf7ff1b30.png)
![Screenshot (192)](https://user-images.githubusercontent.com/69470381/122429995-d7ffee80-cfb0-11eb-932d-9e227c8498ac.png)
![Screenshot (193)](https://user-images.githubusercontent.com/69470381/122430029-df26fc80-cfb0-11eb-9d8f-e9e3ab252e87.png)
![Screenshot (194)](https://user-images.githubusercontent.com/69470381/122430088-eb12be80-cfb0-11eb-8f8a-c3277fbf8209.png)
![Screenshot (195)](https://user-images.githubusercontent.com/69470381/122430143-f4039000-cfb0-11eb-8a23-085d196d4150.png)

**Reply by sfromis** [reply]
That was not the issue. You wrote about trouble like "brick", not just difference in included features. Your 9.5 build just did not have the same features active at build time. AFAICT, your upgrade now was **entirely successful**.

**Reply by NsinghP** [reply]
> That was not the issue. You wrote about trouble like "brick", not just difference in included features. Your 9.5 build just did not have the same features active at build time. AFAICT, your upgrade now was **entirely successful**.

The problem is that if I do Reset 1 at this point, the device will not do anything until a device recovery sequence with 6 ons and offs is initiated. Edit, also the Memory model is changed from 4mb to 1mb

Edit2: Before the OTA upgrade this was the memory info: 

Flash Size 4096 kB
Program Flash Size 4096 kB
Program Size 630 kB
Free Program Space 1416 kB
Free Memory 19.8 kB

Compare that to last picture after upgrade

Edit3: This time I did not have to initiate "device recovery sequence with 6 ons and offs" after the reset 1 command, it simply presented me with Tasmota AP as it should after Reset 1, but my custom name for AP was gone. The point is that though Jason says it is possible to do OTA upgrade on 4MB Memory Model, while maintaining the 4MB memory model, but it looks like that in this case it is not possible to do so!

**Reply by Jason2866** [reply]
You Interpretation is wrong. If you OTA a Tasmota build linked for 1M on a device with 4M flash you will have only 1M to use. Thats normal. Thats why more linker scripts exists. Before this change doing a OTA with a different memory model ended in a softbrick

**Reply by Jason2866** [reply]
Reset1 just did what it should. You have flashed a version WITHOUT your credentials and settings compiled in. So what do you thought will happen if you reset every setting to firmware defaults?

**Reply by NsinghP** [reply]
> You Interpretation is wrong. If you OTA a Tasmota build linked for 1M on a device with 4M flash you will have only 1M to use. Thats normal. Thats why more linker scripts exists. Before this change doing a OTA with a different memory model ended in a softbrick

I may be lacking in my understanding of the whole OTA upgrade process, so please excuse my questions. How do i achieve this linking you mention? Since I am compiling the 4MB model .bin file on tasmocompiler, do I need to add a specific custom string to achieve this linking?

Edit: Let me rephrase the question. Let's say I gift this Tasmota device to my niece. And her boyfriend, since the gentleman he is, decides to upgrade the device, how do I "Fool"proof this device against any unintended upgrade.>Better password is not the solution, because it is assumed that password has been compromised already.

**Reply by Jason2866** [reply]
Build always with Tasmocompiler a 4M build and OTA this firmware and you are fine.
You cant use any precompiled Tasmota firmware since the are build with 1M linker script. When doing this you have 1M you can use
Device will not be bricked. But when you flash a standard Tasmota build (1M)
on a device with a non standard build (4M) you will always end in 1M flash usage

**Reply by barbudor** [reply]
@NsinghP if you want to OTA your own custom build you need to get a webserver (http) of your own and host the bin.gz file (+ tasmota-minimal.bin.gz) and in your `user_config_override.h`, override the default OTA URL to your own like:

```
 #define OTA_URL                "http://myownserver/tasmota/" PROJECT "-"   CODE_IMAGE_STR ".bin.gz"  // [OtaUrl]
```

**Reply by Jason2866** [reply]
Attention the server needs to be **http** https is NOT possible

**Reply by NsinghP** [reply]
> if you want to OTA your own custom build you need to get a webserver (http)

ha ha, at this point, I am thinking that I am better off going two states over and kicking her boyfriends ass! :)

**Reply by NsinghP** [reply]
> Build always with Tasmocompiler a 4M build and OTA this firmware and you are fine.

But I thought I was already building a 4M build with tasmocompiler! What am I missing?

**Reply by sfromis** [reply]
What I saw you post strongly suggested that you replaced your self-compiled binary with standard precompiled tasmota.bin, which obviously does not include whatever you chose for your own compile. When upgrading firmware, it will **never** retain features from the previous binary, but will normally preserve runtime options configured via commands or the web UI. If you then do a ` Reset 1` these retained settings will also be erased.

**Reply by NsinghP** [reply]
> 
> 
> What I saw you post strongly suggested that you replaced your self-compiled binary with standard precompiled tasmota.bin, which obviously does not include whatever you chose for your own compile. When upgrading firmware, it will **never** retain features from the previous binary, but will normally preserve runtime options configured via commands or the web UI. If you then do a ` Reset 1` these retained settings will also be erased.

Okay, that makes sense, but how do i make sure (other than the method suggested by barbudor) that I don't replace my precompiled binary with standard tasmota.bin when doing OTA upgrade by mistake? One obvious solution is a warning next to the "Upgrade by web server" "Start upgrade" button stating that "do not upgrade if using Self-compiled binary".

**Reply by sfromis** [reply]
Well, barbudor already pointed out the perfect solution, no need to improve on that. You could even include a non-working OTA url in your binary. That is a really excellent way of making click-happy fingers need to do more work before anything happens.

**Reply by NsinghP** [reply]
> 
> 
> Well, barbudor already pointed out the perfect solution, no need to improve on that. You could even include a non-working OTA url in your binary. That is a really excellent way of making click-happy fingers need to do more work before anything happens.

agreed, a non working url is an excellent solution. But that still leaves the question: Can standard OTA upgrade work for anykind of self-compiled binary? The expectation being that upgrade will not damage existing setup but only upgrade the core, especially if the upgrade addresses any identified security vulnerabilities.

Edit: It would also be desirable to see an OTA upgrade that calculates what is needed for features to be retained and at the least,  refuse to upgrade, maybe giving a message that current firmware is not compatible with this upgrade.

Edit: Here are some pictures of the device I am making as gift on my niece's Birthday, I have few more months to make it better:

![1](https://user-images.githubusercontent.com/69470381/122451011-51a1d780-cfc5-11eb-91c2-633f2f726afb.jpeg)
![2](https://user-images.githubusercontent.com/69470381/122451048-56ff2200-cfc5-11eb-9004-e3d49f93b06d.jpeg)

**Reply by Jason2866** [reply]
No

**Reply by sfromis** [reply]
It would not be possible for OTA to calculate what features **you** feel are needed when you flash another binary. And it is entirely normal that irresponsible OTA may have rather detrimental effects on the ability to work properly. While those risks are lower with different memory model no longer an issue, there can always be things like incompatible settings layout, or a non-working binary flashed. When you decide to jump, you can't expect a safety net other than good preparation.

**Reply by NsinghP** [reply]
Thanks for your time (everyone) obvious solution is custom non https url, hosting the needed file. I am going to do some research on where this file can be hosted.

**Reply by techman83** [reply]
I just use nginx from my local home automation host, I don't really want my IoT devices to be able to talk to the internet (they are fire walled off from it entirely). In theory you might be able to use a ESP device with a sd card. But I haven't tried that personally.

I can at least grab a bin file from one of my ESP32s that have a SD Card
```bash
--2021-06-18 10:05:02--  http://doorbell/ufsd?download=/tasmota32-doorbell.bin
Resolving doorbell (doorbell)... 10.0.0.101
Connecting to doorbell (doorbell)|10.0.0.101|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1505728 (1.4M) [application/octet-stream]
Saving to: 'firmware.bin’

firmware.bin                                                                                    100%[====================================================================================================================================================================================================================================================>]   1.44M   176KB/s    in 9.6s    

2021-06-18 10:05:11 (153 KB/s) - 'firmware.bin’ saved [1505728/1505728]
```

**Reply by NsinghP** [reply]
> I can at least grab a bin file from one of my ESP32s that have a SD Card

Thanks for your suggestion,I will have to try that. ESP32's are so hard to solder, I already destroyed one trying to make it stick to a breakout adapter. Rest I am trying to keep everything in the cloud as much as possible to avoid having to maintain unnecessary servers at home, for instance I am using HiveMQ.cloud with TLS to connect few devices I am using at home. But looks like there is no way around it, I may end up having to host everything needed on a small box like rasperberry pie and leaving it on 24 hrs.

**Reply by NsinghP** [reply]
> if you want to OTA your own custom build you need to get a webserver (http) of your own and host the bin.gz file (+ tasmota-minimal.bin.gz) and in your `user_config_override.h`, override the default OTA URL to your own like:

Could it be that a magnet link can be used to accomplish the same thing? That way I can upload the file and be able to download it via bit torrent system? That actually would be nice system to use for the official OTA download too, but maybe it will add too much overhead to Tasmota size?

**Reply by benzino77** [reply]
If you are running TasmoCompiler in docker on your home environment, you can specify Tasmocompiler instance itself as OTA server:
http://<ip_of_TasmoCompiler>/download/firmware.bin.gz 

There is no need to setup Nginx or any other solution to achiew what you need. There are two requirements:
1. You have run compilation on TC and you are on the last step (you can see download buttons)
2. You are running TC in location which can be reached by your device (<ip_of_TasmoCompiler>)

**Reply by NsinghP** [reply]
> If you are running TasmoCompiler in docker on your home environment, you can specify Tasmocompiler instance itself as OTA

Thanks, I will try to understand what you are saying and then try it out. But Still I think it won't solve the problem faced by 4M memory model devices when their OTA upgrade button is pressed by a novice or by mistake.

**Reply by benzino77** [reply]
Sure it will not be "novice" immune. It will also not work in case there is "OTA magic" needed - intermediate minimal.bin.gz flush.

**Reply by NsinghP** [reply]
My conclusion:

Even if I compile with tasmocompiler using "Generic" option there is a chance that my device will have problems after OTA upgrade as clearly the official OTA bin file uses program flash size of 1024KB and Tasmocompiler bin file uses program Flash Size of 4MB, but that in itself may not be a problem.

If you are using Card/Little FS option or MQTT-TLS option,(or maybe other options/features that are not included in standard tasmota.bin file)  then you will lose these features after OTA upgrade. Therefore, official OTA  url upgrade is only suited for devices that are running on unmodified precompiled binary downloaded from tasmota site. 

If you are interested in OTA upgrade of your custom binary file then it is best to use your own "http" server to host the custom upgrade binary and update the OTA url accordingly as suggested by @barbudor or do a wired/serial flash or erase/flash as suggested by @sfromis .

Onother option is to use the url "http://<ip_of_TasmoCompiler>/download/firmware.bin.gz" as OTA url, as suggested by @benzino77, but to use this method you need access to final page of tasmocompiler where the file download option is shown and the device being flashed has internet access to be able to reach this <ip_of_TasmoCompiler>. But I believe this option can only be used by slightly technically savvy user and is not for casual end users.

Please correct if some of my assumptions or conclusions are inaccurate.

Regards,

Navi

**Reply by benzino77** [reply]
> Even if I compile with tasmocompiler using "Generic" option there is a chance that my device will have problems after OTA upgrade as clearly the official OTA bin file uses program flash size of 1024KB and Tasmocompiler bin file uses program Flash Size of 4MB, but that in itself may not be a problem.

No. Generic is using the same memory map as official Tasmota bin files. So OTA will work fine but you will probably loose your compiled in features as you suggested below.

> If you are using Card/Little FS option or MQTT-TLS option,(or maybe other options/features that are not included in standard tasmota.bin file) then you will lose these features after OTA upgrade. Therefore, official OTA url upgrade is only suited for devices that are running on unmodified precompiled binary downloaded from tasmota site.

If you choose Generic board in TasmoCompiler the option "Card/Little FS" is unchecked. You can enable it in case you are going to connect SD Card slot to your device. Generic memory map will not create filesystem in chip memory because there is no space for that.
And yes it should be rule of thumb that you are using official OTA url for devices flushed with official Tasmota builds. For other cases you should know what you are doing.

> If you are interested in OTA upgrade of your custom binary file then it is best to use your own "http" server to host the custom upgrade binary and update the OTA url accordingly as suggested by @barbudor or do a wired/serial flash or erase/flash as suggested by @sfromis .

You don't need your OTA server to do an upgrade. You can always choose "Upgrade by file upload" and upload your new custom binary from the GUI. Upgrading by "OTA url" is able to do "OTA magic" - which can be described as:
*if there is no space on device to perform one step upgrade, flush minimal version of Tasmota first and then perform final upgrade*

> Onother option is to use the url "http://<ip_of_TasmoCompiler>/download/firmware.bin.gz" as OTA url, as suggested by @benzino77, but to use this method you need access to final page of tasmocompiler where the file download option is shown and the device being flashed has internet access to be able to reach this <ip_of_TasmoCompiler>. But I believe this option can only be used by slightly technically savvy user and is not for casual end users.

I'm not sure this method is tech savvy :)

**Reply by NsinghP** [reply]
> I'm not sure this method is tech savvy :)

Ha ha, thanks for your detailed reply. and here I was getting all happy that I am so tech savvy!

**Reply by NsinghP** [reply]
> if you want to OTA your own custom build you need to get a webserver (http) of your own and host the bin.gz file (+ tasmota-minimal.bin.gz) and in your `user_config_override.h`, override the default OTA URL to your own like:

 #define OTA_URL                "http://myownserver/tasmota/" PROJECT "-"   CODE_IMAGE_STR ".bin.gz"  // [OtaUrl]

(1) In above custom string do I enter a project name  PROJECT "-"  or  leave it as it is? (I am already using a custom project name by using #define PROJECT string)             

(2) I am assuming in CODE_IMAGE_STR ".bin.gz" I need to enter the exact name of my .bin.gz file for example "Hazel.bin.gz"?

Regards,

Navi

**Reply by sfromis** [reply]
In that example, PROJECT and CODE_IMAGE_STR are references to the #define you already have for those, you effectively get a concatenation of constants and other values. But what you choose to use there depends on what naming standard you want for the stored binaries. If you just want the same always, just a single quoted string will also do.

**Reply by NsinghP** [reply]
>  If you just want the same always, just a single quoted string will also do.

Thanks, Then I should be fine with just using #define OTA_URL  "http://mycustomfiles.com/tasmota/Hazel_9-5-0.bin"  // [OtaUrl]

Just curious about the other two, do I really need them? I am wondering If they help in enabling the "Tasmota Magic" feature? Then I have to worry about and understand the naming standard and accordingly modify the PROJECT "-" CODE_IMAGE_STR ".bin.gz" and upload/host the minimal.bn.gz if needed to complete the "Magic maneuver", or is minimal.bin.gz  needed regardless of whether I use these additional strings or not?

**Reply by barbudor** [reply]
It doesn't sounds logical to me to have the version number in the binary name.
That means that your Hazel_9-5-0.bin can only be upgraded to itself as 9.5.0 ?
It also should be the `.bin.gz` for more efficient OTA

You should have a look at how Tasmota is calculating the file name for the minimal build. With such an OTA name, Tasmota will try to load `Hazel_9-5-minimal.bin` (or `Hazel_9-5-minimal.bin.gz` if you go for the .bin.gz one).
https://github.com/arendst/Tasmota/blob/0d18174bd93c721259c2816d04d84ba5fe9476ae/tasmota/support_tasmota.ino#L1119-L1150

At least, swapping underscore and dashes such as `Hazel-9_5_0.bin.gz` would make Tasmota search for `Hazel-minimal.bin.gz`

**Reply by NsinghP** [reply]
Thanks, I will have to try it out practically and then reverse engineer my way to fully understand what you just explained.

Edit: I decided to use non workding url for the time being. I think now I have better understanding of what your command does. Here is what I used: 

#define OTA_URL "http://confirm-and-enter-http-url" PROJECT "-"  CODE_IMAGE_STR ".bin.gz"  // 

and now as a result of above define my OTA url field shows: 

url: http://confirm-and-enter-http-urlHazel-TasmoCompiler-esp82664M.bin.gz

I think with this the device is pretty much safe from "Novice" initiated OTA upgrade.

Now I also understand what your command does: It picks up the project name and file name from the binary to populate the url.

**Comment by benzino77** [comment, marked as answer]
For Wemos/NodeMCU boards, linker memory map is set, by defaullt, to 2MB for program and 2MB for data. Based on what Adrian wrote above, you can do your calculations, but basically you can have pretty big Tasmota image on those boards and still no worries about OTA (900KB firmware and still 900KB for OTA).


2. Have a look [here](https://github.com/benzino77/tasmocompiler/blob/master/FEATURES_DESC.md)
3. I don't think It is "doable". Features change from Tasmota version to Tasmota version. The same feature on version 9.3.1 can be much "lighter" then on 9.3.3 version (just an example). On the other hand TasmoCompiler is grouping sensors/features into groups (just to make it simpler). Those groups can be expanded in feature releses of TasmoCompiler when they appear in Tasmota. 

Some information about the feature size can be found in the Tasmota itself. Look for example [here](https://github.com/arendst/Tasmota/blob/009ea7156c61845ff42cfdb4406a157d3ad4bd4d/tasmota/my_user_config.h#L411)

PS.
Sensors/features from the group, let's say "Light Sensors", can be removed by "Custom parameters" text field:
```C
#undef USE_VEML6075
```
and in the final Tasmota image all "Light sensors" will be included except VEML6075

**Reply by NsinghP** [reply]
> For Wemos/NodeMCU boards, linker memory map is set, by defaullt, to 2MB for program and 2MB for data.

Is this applicable to bare ESP8266 12F chip too or is this a feature that wemos/nodMCU add outside the chip?

**Reply by benzino77** [reply]
You have to remeber that those little boards has limitted amount of RAM, so you probably won't be able to build image with lots of features enabled - read [this](https://github.com/benzino77/tasmocompiler/issues/86) discussion.

**Reply by NsinghP** [reply]
> ou have to remeber that those little boards has limitted amount of RAM, so you probably won't be able to build image with lots of features enabled - read [this](https://github.com/benzino77/tasmocompiler/issues/86) discussion.

Ok, Thanks,

**Reply by benzino77** [reply]
It is applicable to boards "NodeMCU like" (4MB boards). This memory map is applied when you choose "Wemos/NodeMCU" board.

**Reply by NsinghP** [reply]
> It is applicable to boards "NodeMCU like" (4MB boards). This memory map is applied when you choose "Wemos/NodeMCU" board.

That explains why I was having OTA upgrade problems when using Generic Option!

**Reply by Jason2866** [reply]
Tasmota provides different flash size configs for ESP8266 / ESP32.
All parameters which are used are provided via `boards` definition.
You find all supported build variants in the root of the project in folder `boards`.
Other build flash configurations are not supported. Can work, you walk alone if you do different.
I defined own boards definitions because it is the easiest way to take care of the needs of Tasmota and keep platformio setup "clean"

**Reply by benzino77** [reply]
I can add this information (that it applies to all 4MB boards) to the tooltip displayed when you move mouse cursor on "Wemos/NodeMCU" option.

**Reply by Jason2866** [reply]
I reduced the flash size for the ESP8266 from 1M to the max. Tasmota code size
`"maximum_size": 995326,` By doing this a compile/linker error should happen when over this border.

**Reply by benzino77** [reply]
@Jason2866 I will have to make some changes to TC after 9.5 release. I think it will make board config much simpler from TC perspective. Good move I think.

**Reply by Jason2866** [reply]
Another benefit you see all relevant infos during platformio run.
```
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/esp8266_2M256.html
PLATFORM: Espressif 8266 (2.6.3) > Espressif Generic ESP8266 Tasmota 1M sketch 772k OTA 256k SPIFFS
HARDWARE: ESP8266 160MHz, 80KB RAM, 972.00KB Flash
```

**Reply by Jason2866** [reply]
@benzino77 Release 9.5 is planned for tomorrow. Thats old news...
After the release we will use a customized Arduino Esp32 core (as we already do for the ESP8266)
We decided to do since we have new features which needs changes in sdkconfig of the IDF.
This is resulting in different precompiled libs for Arduino ESP32.
The biggest benefit, needed flash space for standard Tasmota32 is reduced about 300 kbytes

**Reply by benzino77** [reply]
@Jason2866 so this is now version 15 or is it a mistake?

**Reply by Jason2866** [reply]
It is 9.5. We use autotagging for releasing via github actions. It is less work to change the (wrong auto-) tag than doing a correct tagged release...
Doing this way releasing is just a push to branch master

**Reply by benzino77** [reply]
@Jason2866 I'm loving it! Moving most of the specific configuration to board files is a good move. It is making my TC configuration simpler and more resistant for Tasmota changes.

Suggestion:
Zigbee Bridge is using board `esp8266_2M256`. How about addind another board, let's say `zbbridge_2M256` which will include `f_cpu: 160000000L`?

**Reply by Jason2866** [reply]
I was considering this too. Decided against because it will end in many boards definitions which differs only in speed. I have to find a user friendly solution to change MCU clock individual for a specific board. I removed the way it was done before because that changed the speed global for all build variants.
The "speed" thing is No.1 on my Tasmota todo list

**Reply by Jason2866** [reply]
@benzino77 
I had to make a compromise between complexity and user friendlieness.
Override is now separated with #12402 for esp8266 / esp32.
I will do a extra boards definition for the zbbridge after this PR is merged
So i think you are on the safe side with Tasmocompiler
